malloc_instrumentation.c

```diff
@@ -0,0 +1,79 @@
+#include <assert.h>
```

Comment:
I think this implementation of malloc/free/realloc should be merged with the one in `basic_testing.h` right? I added this file just for showing the idea of instrumenting the malloc